kafka: seek to correct start offset when creating kafka queue

aljoscha opened the pull request:
Before, we would always start a new queue from Offset:Beginning and then
later seek the consumer when noticing that the offsets we're getting are
not what we expect.

Now, we seek to the correct start offset when creating the queue for a
new partition.

Fixes #9101
Fixes #8900

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [N/A] This PR adds a release note for any
  [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

ruchirK commented on test/testdrive/kafka-time-offset.td:
```diff
@@ -304,3 +304,37 @@ cherry    3
+
+> CREATE MATERIALIZED SOURCE verify_no_fetch
+  FROM KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'testdrive-t4-${testdrive.seed}'
+  WITH (kafka_time_offset = -1)
```
this test would be slightly more clear for me if we could specify `start_offsets` directly, or maybe with a comment to indicate why that's not possible.

aljoscha replied:
The glib answer would be: because I'm testing the `kafka_time_offset` feature. The better answer is that I came to the bug from that angle but will now also add a test that uses `start_offset` in `kafka-avro-sources.td` because you made me realize that I should. :sweat_smile:  That duplicates testing a bit, because the same mechanism is used underneath, but maybe that's okay. What do you think @philip-stoev?

I'll also add a comment here, saying that a time offset of `-1` means we start from the end.

philip-stoev replied:
The more tests, the better. I personally have nothing against duplication in testing -- no two tests are alike even if one thinks they are, so they will usually hit different issues.

ruchirK reviewed (approved):
LGTM!

philip-stoev reviewed (approved):
The particular issue, `Fetch topic ... at offset 0` is gone, as evidenced by your test.

That said, with persistent-kafka-upsert-source, on restart Mz still performs `Fetch topic testdrive-s1-30 [0] at offset N` where `N` is no longer zero but is still smaller than the desired number. So if you could merge this PR in, I will be able to open the next ticket in this series.

aljoscha commented:
This change introduces an instability. When creating the consumer queues with a specific `Offset`, instead of `Offset::Beginning` we seem to provoke rdkafka into adding messages to the main consumer queue, which we don't expect and leads to lost messages. This could be related to https://github.com/edenhill/librdkafka/issues/3231.

The original issue this is trying to fix is "only" an inefficiency in that we're reading messages that we're then throwing away and I haven't managed to find the reason for the instability that this change introduces yet. I'm somewhat inclined to drop the issue for now.

philip-stoev commented:
@aljoscha the only problem with dropping this issue now is that it will not be possible to verify in tests that persistence has kicked in at all.

Tests would not be able to use a check such as this one post-restart to make sure that persistence indeed prevented re-ingestion of everything:

```
SELECT SUM((statistics->'topics'->'testdrive-no-fetch-test-${testdrive.seed}'->'partitions'->'0'->'msgs')::int) = 0 from mz_kafka_source_statistics;
```

Therefore, it is possible for persistence to fail to kick in entirely due to some regression and all the tests will happily continue to pass since all the data will be ingested from scratch post-restart and the source will look and feel the same as if persistence had kicked in.

aljoscha commented:
Good news, everyone! I found the source of the instability that appears when setting an initial partition offset other than `Offset::Beginning`. Turns out, that assigning partitions/topics and splitting off individual partition queues is not atomic (which doesn't seem too surprising for me, now that I think about it). I added a test case that highlights this behavior: https://github.com/MaterializeInc/materialize/pull/9102/files#diff-ad5d38bf37b32dbbd4546b0fcd7f9b6daf573be8d5392847e1ed39c09b6c9f5cR656.

I'm fixing this by handling messages that we get from `consumer.poll()`. Before we we're treating these as errors because we were (wrongly) assuming that no messages would ever show up. Additionally, I have refactored `get_next_message()` to make it a bit easier to understand.

cc @ruchirK , @petrosagg (you might be interested), @elindsey (maybe you have an opinion because you wrote the original code that errors on `consumer.poll()`)

philip-stoev commented:
I called it! On Apr 27th! In https://github.com/MaterializeInc/materialize/issues/6586:

>  We have at least one instance where we call poll() and then confidently discard the message.

BRB as I do a little victory dance.

aljoscha commented:
:sweat_smile: 

Can you see if it fixes that one as well? In my reproducing test, the instability only showed up after the other change that I did in this PR. When not seeking to an initial offset via `Offset::Offset(o)` it doesn't show. So that other bug might be something else.

aljoscha commented:
Also: congratulations! :1st_place_medal:

aljoscha commented:
@ruchirK Unfortunately, I don't think I can make it more precise than what I have already, i.e. "we expect some messages" (might be none, or one, or many). :see_no_evil: 

I updated the reproducer test to observe how many messages we get on the common queue and the partition queue, and to print that on failure. We most often get `0` elements in the common queue, so it takes a while to observe a failure. Sometimes we get `4` out of `1000`, sometimes `1000` out of `1000`. It depends on the record size, how many bytes the fetcher thread (I guess) fetches per request (I configured that) and maybe other things.

aljoscha commented:
One thing is clear though: once the switch to the partition queue happened we don't get any more message on the consumer queue. I added a check for that in the test and it never tripped.

ruchirK commented:
Ok cool! this change seems fine to merge to me
